### PR TITLE
Fixed Gemma example using Gemma2 models

### LINKED ISSUE #4567: Some surprises when adding Gemma2-2b to flax/examples/gemma
Looking forward to getting to grips with `nnx`, I went about implementing the apparently missing `gemma2_2b` model.

After coming up with a suitable `TransformerConfig`, I was surprised that I couldn't get a forward pass to run : So I added 'Grouped-Query Attention' (mirroring what can be seen in the GDM `gemma` library).  

Still no dice, since (although I checked all the params were getting set correctly), garbage logits were being returned.  So I labouriously went through each layer, checking for numerical differences, only to find that in [`modules.Block`](https://github.com/google/flax/blob/main/examples/gemma/modules.py#L338-L358), the ordering of the sub-modules is clearly in the wrong order (look at the `pre_ffw_norm` and the `post_attn_norm`, and the residual paths, for instance):

```python
class Block(nnx.Module):
    #...
  def __call__(self, x, ...):  
    inputs_normalized = self.pre_attention_norm(x)
    cache, attn_output = self.attn(
        inputs_normalized,
        segment_pos,
        cache,
        attn_mask,
    )
    attn_output += x
    residual = attn_output
    attn_output = self.pre_ffw_norm(attn_output)

    if self.use_post_attn_norm:
      attn_output = self.post_attn_norm(attn_output)
    self.sow_config.maybe_sow_rs_after_attention(attn_output, self)

    outputs = self.mlp(attn_output)
    if self.use_post_ffw_norm:
      outputs = self.post_ffw_norm(outputs)
    outputs = residual + outputs
    self.sow_config.maybe_sow_rs_after_ffw(outputs, self)
    return cache, outputs
```

After fixing this up, the model's final `logits` come out pretty close to the GDM `gemma` ones (small errors seem to creep in along the way due to bfloat16/float16).  This is good news - though it is slightly worrying that this example has been *unusable* over the whole period, since its actual functionality as an implementation of gemma models does not appear to have been tested...

Please let me know if you'd like a PR to cover (at minimum) adding a working `gemma2_2b`

### PR DESCRIPTION
**Pull request by vfdev-5**
# What does this PR do?

This PR is based on https://github.com/google/flax/pull/4587 (thanks to @mdda!) and adds more tests and fixes remaining incosistencies with the latest codebase.

Fixes #4567
Closes #4587

## Checklist
- [x] This PR fixes a minor issue (e.g.: typo or small bug) or improves the docs (you can dismiss the other
      checks if that's the case).
- [ ] This change is discussed in a Github issue/
      [discussion](https://github.com/google/flax/discussions) (please add a
      link).
- [ ] The documentation and docstrings adhere to the
      [documentation guidelines](https://github.com/google/flax/blob/main/docs/README.md#how-to-write-code-documentation).
- [ ] This change includes necessary high-coverage tests.
      (No quality testing = no merge!)

### CODE REVIEW

File: examples/gemma/layers.py [hunk 1]
```diff
@@ -51,7 +52,7 @@ def __init__(self, dim: int, *, rngs: nnx.Rngs):
 
   def __call__(self, x: Array) -> Array:
     var = jnp.mean(jnp.square(x), axis=-1, keepdims=True)
-    normed_inputs = jnp.asarray(x * jnp.reciprocal(jnp.sqrt(var + 1e-06)))
+    normed_inputs = jnp.asarray(x * jax.lax.rsqrt(var + 1e-06))
```
Comment (vfdev-5) [inline examples/gemma/layers.py]: This what also MaxText is doing: https://github.com/AI-Hypercomputer/maxtext/blob/33339821e96327d25613ace9d28c7a8789698586/MaxText/layers/normalizations.py#L60

File: examples/gemma/sampler_test.py [hunk 2]
```diff
@@ -519,6 +520,51 @@ def test_compute_attention_mask(self):
 
     self.assertTrue((attn_mask.squeeze(1) == expected_attn_mask).all())
 
+  @parameterized.parameters(
+    {"url": "google/gemma/flax/2b"},
+    {"url": "google/gemma-2/flax/gemma2-2b"},
+    {"url": "google/gemma-3/flax/gemma3-1b"},
+  )
+  def test_models_from_kaggle(self, url):
```
Comment (vfdev-5) [inline examples/gemma/sampler_test.py:528]: @cgarciae I added this test to make sure that real models can work. By default this test is skipped and it can run locally if kagglehub is installed and env vars have creds.
Let me know if we want / do not want to keep this.

**Review by cgarciae — APPROVED** [review]

**Comment by cgarciae** [comment]
Thanks @vfdev-5 (and @mdda)!
